builtin/providers/vsphere/resource_vsphere_virtual_machine_test.go

```diff
@@ -127,6 +127,67 @@ func TestAccVSphereVirtualMachine_dhcp(t *testing.T) {
 	})
 }
 
+func TestAccVSphereVirtualMachine_custom_configs(t *testing.T) {
+	var vm virtualMachine
+	var locationOpt string
+	var datastoreOpt string
+
+	if v := os.Getenv("VSPHERE_DATACENTER"); v != "" {
+		locationOpt += fmt.Sprintf("    datacenter = \"%s\"\n", v)
+	}
+	if v := os.Getenv("VSPHERE_CLUSTER"); v != "" {
+		locationOpt += fmt.Sprintf("    cluster = \"%s\"\n", v)
+	}
+	if v := os.Getenv("VSPHERE_RESOURCE_POOL"); v != "" {
+		locationOpt += fmt.Sprintf("    resource_pool = \"%s\"\n", v)
+	}
+	if v := os.Getenv("VSPHERE_DATASTORE"); v != "" {
+		datastoreOpt = fmt.Sprintf("        datastore = \"%s\"\n", v)
+	}
+	template := os.Getenv("VSPHERE_TEMPLATE")
+	label := os.Getenv("VSPHERE_NETWORK_LABEL_DHCP")
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckVSphereVirtualMachineDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: fmt.Sprintf(
+					testAccCheckVSphereVirtualMachineConfig_custom_configs,
+					locationOpt,
+					label,
+					datastoreOpt,
+					template,
+				),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckVSphereVirtualMachineExists("vsphere_virtual_machine.car", &vm),
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "name", "terraform-test-custom"),
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "vcpu", "2"),
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "memory", "4096"),
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "disk.#", "1"),
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "disk.0.template", template),
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "network_interface.#", "1"),
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "custom_configuration_parameters.foo", "bar"),
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "custom_configuration_parameters.car", "ferrai"),
+					resource.TestCheckResourceAttr(
+						"vsphere_virtual_machine.car", "custom_configuration_parameters.num", "42"),
```

Comment:
These helpers are only checking the state file, so right now no verification is being done on the custom parameters.

In other words, if you remove all your code but the addition of the field to the schema, this test will still pass!

Something along these lines:

```go
func testAccCheckVSphereVirtualMachineCustomConfigs(vm *virtualMachine) resource.TestCheckFunc {
  return func(s *terraform.State) error {
    // use the vmware client to ensure that the passed in vm has the expected custom config
    // return an error if not
  }
}
```

I'd recommend writing a `CheckFunc` that accepts `&vm` and inspects it to prove the custom attributes get se the